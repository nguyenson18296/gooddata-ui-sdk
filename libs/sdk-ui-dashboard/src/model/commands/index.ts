// (C) 2021 GoodData Corporation

import { LoadDashboard, RenameDashboard, ResetDashboard, SaveDashboard, SaveDashboardAs } from "./dashboard";
import {
    AddAttributeFilter,
    ChangeAttributeFilterSelection,
    ChangeDateFilterSelection,
    MoveAttributeFilter,
    RemoveAttributeFilters,
    SetAttributeFilterParent,
} from "./filters";
import {
    ChangeInsightWidgetFilterSettings,
    ChangeInsightWidgetHeader,
    ChangeInsightWidgetInsight,
    ChangeInsightWidgetVisProperties,
    ModifyDrillsForInsightWidget,
    RefreshInsightWidget,
    RemoveDrillsForInsightWidget,
} from "./insight";
import {
    ChangeKpiWidgetComparison,
    ChangeKpiWidgetFilterSettings,
    ChangeKpiWidgetHeader,
    ChangeKpiWidgetMeasure,
    RefreshKpiWidget,
} from "./kpi";
import {
    AddLayoutSection,
    AddSectionItems,
    ChangeLayoutSectionHeader,
    MoveLayoutSection,
    MoveSectionItem,
    RemoveLayoutSection,
    RemoveSectionItem,
    ReplaceSectionItem,
    UndoLayoutChanges,
} from "./layout";
import { CreateAlert, RemoveAlert, UpdateAlert } from "./alerts";
import { CreateScheduledEmail } from "./scheduledEmail";
import {
    Drill,
    DrillDown,
    DrillToAttributeUrl,
    DrillToCustomUrl,
    DrillToDashboard,
    DrillToInsight,
    DrillToLegacyDashboard,
} from "./drill";

export { DashboardCommandType, IDashboardCommand, CommandProcessingMeta } from "./base";
export {
    InitialLoadCorrelationId,
    LoadDashboard,
    loadDashboard,
    SaveDashboardAs,
    saveDashboardAs,
    SaveDashboard,
    saveDashboard,
    RenameDashboard,
    renameDashboard,
    ResetDashboard,
    resetDashboard,
} from "./dashboard";
export {
    ChangeDateFilterSelection,
    changeDateFilterSelection,
    clearDateFilterSelection,
    AddAttributeFilter,
    addAttributeFilter,
    MoveAttributeFilter,
    moveAttributeFilter,
    RemoveAttributeFilters,
    removeAttributeFilter,
    ChangeAttributeFilterSelection,
    AttributeFilterSelectionType,
    resetAttributeFilterSelection,
    changeAttributeFilterSelection,
    SetAttributeFilterParent,
    setAttributeFilterParent,
} from "./filters";
export {
    AddLayoutSection,
    addLayoutSection,
    MoveLayoutSection,
    moveLayoutSection,
    RemoveLayoutSection,
    removeLayoutSection,
    ChangeLayoutSectionHeader,
    changeLayoutSectionHeader,
    AddSectionItems,
    addSectionItem,
    ReplaceSectionItem,
    replaceSectionItem,
    MoveSectionItem,
    moveSectionItem,
    RemoveSectionItem,
    eagerRemoveSectionItem,
    UndoLayoutChanges,
    undoLayoutChanges,
    revertLastLayoutChange,
    DashboardLayoutCommands,
    UndoPointSelector,
} from "./layout";
export { CreateAlert, createAlert, RemoveAlert, removeAlert, UpdateAlert, updateAlert } from "./alerts";
export { CreateScheduledEmail } from "./scheduledEmail";
export {
    Drill,
    DrillDown,
    DrillToAttributeUrl,
    DrillToCustomUrl,
    DrillToDashboard,
    DrillToInsight,
    DrillToLegacyDashboard,
    drill,
    drillDown,
    drillToAttributeUrl,
    drillToCustomUrl,
    drillToDashboard,
    drillToInsight,
    drillToLegacyDashboard,
} from "./drill";
export {
    ChangeKpiWidgetHeader,
    changeKpiWidgetHeader,
    ChangeKpiWidgetMeasure,
    changeKpiWidgetMeasure,
    ChangeKpiWidgetFilterSettings,
    changeKpiWidgetFilterSettings,
    ChangeKpiWidgetComparison,
    changeKpiWidgetComparison,
    RefreshKpiWidget,
    refreshKpiWidget,
    KpiWidgetComparison,
} from "./kpi";

export {
    ChangeInsightWidgetHeader,
    changeInsightWidgetHeader,
    ChangeInsightWidgetFilterSettings,
    changeInsightWidgetFilterSettings,
    ChangeInsightWidgetVisProperties,
    changeInsightWidgetVisProperties,
    ChangeInsightWidgetInsight,
    changeInsightWidgetInsight,
    ModifyDrillsForInsightWidget,
    modifyDrillForInsightWidget,
    RemoveDrillsForInsightWidget,
    removeDrillForInsightWidget,
    RemoveDrillsSelector,
    RefreshInsightWidget,
    refreshInsightWidget,
} from "./insight";

/**
 * @internal
 */
export type DashboardCommands =
    | LoadDashboard
    | SaveDashboard
    | SaveDashboardAs
    | RenameDashboard
    | ResetDashboard
    | ChangeDateFilterSelection
    | AddAttributeFilter
    | RemoveAttributeFilters
    | MoveAttributeFilter
    | ChangeAttributeFilterSelection
    | SetAttributeFilterParent
    | AddLayoutSection
    | MoveLayoutSection
    | RemoveLayoutSection
    | ChangeLayoutSectionHeader
    | AddSectionItems
    | ReplaceSectionItem
    | MoveSectionItem
    | RemoveSectionItem
    | UndoLayoutChanges
    | ChangeKpiWidgetHeader
    | ChangeKpiWidgetMeasure
    | ChangeKpiWidgetFilterSettings
    | ChangeKpiWidgetComparison
    | RefreshKpiWidget
    | ChangeInsightWidgetHeader
    | ChangeInsightWidgetFilterSettings
    | ChangeInsightWidgetVisProperties
    | ChangeInsightWidgetInsight
    | ModifyDrillsForInsightWidget
    | RemoveDrillsForInsightWidget
    | RefreshInsightWidget
    | CreateAlert
    | UpdateAlert
    | RemoveAlert
    | CreateScheduledEmail
    | Drill
    | DrillDown
    | DrillToAttributeUrl
    | DrillToCustomUrl
    | DrillToDashboard
    | DrillToInsight
    | DrillToLegacyDashboard;
