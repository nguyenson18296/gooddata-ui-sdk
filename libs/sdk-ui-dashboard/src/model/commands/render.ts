// (C) 2021 GoodData Corporation
import { IDashboardCommand } from "./base";

/**
 * @alpha
 */
export interface RequestAsyncRender extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.RENDER.ASYNC.REQUEST";
    readonly payload: {
        /**
         * Async render identifier (eg stringified widget {@link @gooddata/sdk-model#ObjRef}).
         */
        readonly id: string;
    };
}

/**
 * Notify the dashboard about async rendering of the component.
 *
 * Mechanism is following:
 * - You must request async rendering for at least 1 component within 2 seconds of the {@link DashboardLoaded} event.
 *   (If you do not register any asynchronous rendering, after 2 seconds the dashboard will announce that it is rendered.)
 * - You can request async rendering for any number of components. Requests are valid if the first rule is met
 *   and not all asynchronous renderings have been resolved and the maximum timeout (60s by default) has not elapsed.
 * - The component may again request asynchronous rendering within 2 seconds of resolution. Maximum 3x.
 *   (this is necessary to cover re-renders caused by data received from the component - eg pushData)
 *
 * - Each component on the dashboard that is rendered asynchronously should fire this command.
 * - Once the component is rendered, it should notify the dashboard by dispatching {@link resolveAsyncRender} command
 *   with the corresponding identifier.
 *
 * In this way, the dashboard is able to recognize and notify that it is fully rendered.
 * (which is useful, for example, when exporting)
 *
 * @alpha
 * @param id - async render identifier
 * @param correlationId - optionally specify correlation id to use for this command.
 *                        this will be included in all events that will be emitted during the command processing
 * @returns
 */
export function requestAsyncRender(id: string, correlationId?: string): RequestAsyncRender {
    return {
        type: "GDC.DASH/CMD.RENDER.ASYNC.REQUEST",
        correlationId,
        payload: {
            id,
        },
    };
}

/**
 * @alpha
 */
export interface ResolveAsyncRender extends IDashboardCommand {
    readonly type: "GDC.DASH/CMD.RENDER.ASYNC.RESOLVE";
    readonly payload: {
        /**
         * Async render identifier (eg stringified widget {@link @gooddata/sdk-model#ObjRef}).
         */
        readonly id: string;
    };
}

/**
 * Notify the dashboard about resolved async rendering of the component.
 *
 * - Each component on the dashboard that is rendered asynchronously should fire this command.
 * - This command should only be dispatched if a {@link requestAsyncRender} command with the corresponding identifier
 *   has already been dispatched.
 *
 * @alpha
 * @param id - async render identifier
 * @param correlationId - optionally specify correlation id to use for this command.
 *                        this will be included in all events that will be emitted during the command processing
 * @returns
 */
export function resolveAsyncRender(id: string, correlationId?: string): ResolveAsyncRender {
    return {
        type: "GDC.DASH/CMD.RENDER.ASYNC.RESOLVE",
        correlationId,
        payload: {
            id,
        },
    };
}
