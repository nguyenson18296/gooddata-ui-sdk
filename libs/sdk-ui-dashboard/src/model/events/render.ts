// (C) 2021 GoodData Corporation
import { DashboardContext } from "../types/commonTypes";
import { IDashboardEvent } from "./base";

/**
 * This event is emitted as soon as the dashboard component is mounted,
 * and rendering of its content started.
 *
 * @alpha
 */
export interface DashboardRenderRequested extends IDashboardEvent {
    readonly type: "GDC.DASH/EVT.RENDER.REQUESTED";
}

/**
 * @alpha
 */
export function renderRequested(ctx: DashboardContext, correlationId?: string): DashboardRenderRequested {
    return {
        type: "GDC.DASH/EVT.RENDER.REQUESTED",
        correlationId,
        ctx,
    };
}

/**
 * This event is emitted when a component on the dashboard requests async rendering.
 *
 * @alpha
 */
export interface DashboardAsyncRenderRequested extends IDashboardEvent {
    readonly type: "GDC.DASH/EVT.RENDER.ASYNC.REQUESTED";
    readonly payload: {
        /**
         * Item identifier.
         */
        readonly id: string;
    };
}

/**
 * @alpha
 */
export function asyncRenderRequested(
    id: string,
    ctx: DashboardContext,
    correlationId?: string,
): DashboardAsyncRenderRequested {
    return {
        type: "GDC.DASH/EVT.RENDER.ASYNC.REQUESTED",
        correlationId,
        ctx,
        payload: {
            id,
        },
    };
}

/**
 * This event is emitted when a component on the dashboard resolves async rendering.
 *
 * @alpha
 */
export interface DashboardAsyncRenderResolved extends IDashboardEvent {
    readonly type: "GDC.DASH/EVT.RENDER.ASYNC.RESOLVED";
    readonly payload: {
        /**
         * Item identifier.
         */
        readonly id: string;
    };
}

/**
 * @alpha
 */
export function asyncRenderResolved(
    id: string,
    ctx: DashboardContext,
    correlationId?: string,
): DashboardAsyncRenderResolved {
    return {
        type: "GDC.DASH/EVT.RENDER.ASYNC.RESOLVED",
        correlationId,
        ctx,
        payload: {
            id,
        },
    };
}

/**
 * @alpha
 */
export interface DashboardRenderResolved extends IDashboardEvent {
    readonly type: "GDC.DASH/EVT.RENDER.RESOLVED";
}

/**
 * This event is emitted as soon as the dashboard component is fully rendered,
 * and asynchronous rendering of each component is complete.
 *
 * @alpha
 */
export function renderResolved(ctx: DashboardContext, correlationId?: string): DashboardRenderResolved {
    return {
        type: "GDC.DASH/EVT.RENDER.RESOLVED",
        correlationId,
        ctx,
    };
}
