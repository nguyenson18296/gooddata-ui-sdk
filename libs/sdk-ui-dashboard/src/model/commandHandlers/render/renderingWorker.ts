// (C) 2021 GoodData Corporation
import { Task, SagaIterator } from "redux-saga";
import { put, delay, take, join, race, call, all, spawn, cancel, actionChannel } from "redux-saga/effects";
import { DashboardContext } from "../../types/commonTypes";
import { newDashboardEventPredicate, renderResolved, renderRequested } from "../../events";

function* wait(ms: number): SagaIterator<true> {
    yield delay(ms);
    return true;
}

const isAsyncRenderResolvedEvent = (id?: string) =>
    newDashboardEventPredicate(
        "GDC.DASH/EVT.RENDER.ASYNC.RESOLVED",
        id ? (e) => e.payload.id === id : () => true,
    );

const isAsyncRenderRequestedEvent = (id?: string) =>
    newDashboardEventPredicate(
        "GDC.DASH/EVT.RENDER.ASYNC.REQUESTED",
        id ? (e) => e.payload.id === id : () => true,
    );

/**
 * @internal
 */
export interface RenderingWorkerConfiguration {
    /**
     * Maximum time limit for rendering the dashboard.
     *
     * Default: 60000 (60s).
     */
    maxTimeout: number;

    /**
     * Maximum time limit for the first asynchronous rendering request.
     * If no asynchronous rendering request is fired in this time limit, the dashboard will announce that it is rendered.
     *
     * Default: 2000 (2s).
     */
    asyncRenderRequestedTimeout: number;

    /**
     * Maximum time limit to re-request asynchronous rendering of the component once it's resolved.
     *
     * Default: 2000 (2s).
     */
    asyncRenderResolvedTimeout: number;
}

export function newRenderingWorker(
    config: RenderingWorkerConfiguration = {
        asyncRenderRequestedTimeout: 2000,
        asyncRenderResolvedTimeout: 2000,
        maxTimeout: 60000,
    },
) {
    return function* renderingWorker(ctx: DashboardContext): SagaIterator<void> {
        try {
            // First, notify that the rendering of the whole dashboard started.
            yield put(renderRequested(ctx));

            // Wait for the dashboard initialization.
            yield take("GDC.DASH/EVT.LOADED");

            // Collect async rendering requests.
            const asyncRenderTasks = yield call(collectAsyncRenderTasks, config);

            // Wait for the resolution of all async rendering tasks.
            yield call(waitForAsyncRenderTasksResolution, asyncRenderTasks, config);

            // Notify that the dashboard is fully rendered.
            yield put(renderResolved(ctx));
        } catch (err) {
            // eslint-disable-next-line no-console
            console.error("Rendering worker failed", err);
        }
    };
}

function* collectAsyncRenderTasks(config: RenderingWorkerConfiguration): SagaIterator<Map<string, Task>> {
    const asyncRenderTasks = new Map<string, Task>();
    const timeout: Task = yield spawn(wait, config.asyncRenderRequestedTimeout);
    const renderRequestedChannel = yield actionChannel(isAsyncRenderRequestedEvent());

    while (true) {
        const { asyncRenderRequested, timeoutResolved } = yield race({
            asyncRenderRequested: take(renderRequestedChannel),
            timeoutResolved: join(timeout),
        });

        if (timeoutResolved) {
            break;
        }

        const asyncRenderId = asyncRenderRequested.payload.id;

        // Check whether async render is already registered.
        // If so, skip it, because possible re-executions are covered in the async render task itself.
        if (asyncRenderTasks.has(asyncRenderId)) {
            continue;
        }

        // New async render requested, spawn it.
        const asyncRenderTask = yield spawn(waitForAsyncRenderResolution, asyncRenderId, config);
        asyncRenderTasks.set(asyncRenderId, asyncRenderTask);
    }

    return asyncRenderTasks;
}

function* waitForAsyncRenderTasksResolution(
    asyncRenderTasks: Map<string, Task>,
    config: RenderingWorkerConfiguration,
): SagaIterator<void> {
    const timeout: Task = yield spawn(wait, config.maxTimeout - config.asyncRenderRequestedTimeout);
    const renderRequestedChannel = yield actionChannel(isAsyncRenderRequestedEvent());
    while (true) {
        // Now, wait for resolution of all the requested async renders, or max 60 seconds.
        // Accept also new async render requests, but only in case that all tasks are not already resolved.
        const asyncRenderTasksList = Array.from(asyncRenderTasks, ([, task]) => task);

        const { asyncTasksResolved, asyncRenderRequested, timeoutResolved } = yield race({
            asyncRenderRequested: take(renderRequestedChannel),
            asyncTasksResolved: all(asyncRenderTasksList.map((task) => join(task))),
            timeoutResolved: join(timeout),
        });

        if (timeoutResolved || asyncTasksResolved) {
            asyncRenderTasksList.forEach((task) => task.cancel());
            yield cancel(timeout);
            return;
        }

        // If the previous async rendering tasks have not yet been resolved,
        // also accept the new asynchronous rendering requests.
        if (asyncRenderRequested) {
            const asyncRenderId = asyncRenderRequested.payload.id;

            // Check whether async render is already registered.
            // If so, skip it, because possible re-executions are covered in the async render task itself.
            if (asyncRenderTasks.has(asyncRenderId)) {
                continue;
            }

            // New async render requested, track it.
            const asyncRenderTask = yield spawn(waitForAsyncRenderResolution, asyncRenderId, config);
            asyncRenderTasks.set(asyncRenderId, asyncRenderTask);
        }
    }
}

function* waitForAsyncRenderResolution(id: string, config: RenderingWorkerConfiguration): SagaIterator<void> {
    // Avoid infinite loop
    const maxRetries = 3;
    let retries = 0;
    const renderResolvedChannel = yield actionChannel(isAsyncRenderResolvedEvent(id));
    const renderRequestedChannel = yield actionChannel(isAsyncRenderRequestedEvent(id));

    while (true) {
        yield take(renderResolvedChannel);

        // Wait for possible re-execution to cover executions
        // invoked by changed execution definition after visualization render.
        // (e.g. by data received from PluggableVisualization pushData callback)
        const { timeout } = yield race({
            timeout: call(wait, config.asyncRenderResolvedTimeout),
            anotherExecution: take(renderRequestedChannel),
        });

        retries += 1;

        // No re-execution, or max retry limit reached - leave the loop
        if (timeout || retries === maxRetries) {
            return;
        }
    }
}
