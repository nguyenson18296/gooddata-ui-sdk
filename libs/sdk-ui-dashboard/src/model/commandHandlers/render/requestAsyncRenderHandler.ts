// (C) 2021 GoodData Corporation
import { DashboardContext } from "../../types/commonTypes";
import { internalErrorOccurred } from "../../events/general";
import { RequestAsyncRender } from "../../commands/render";
import { asyncRenderRequested } from "../../events/render";

export function requestAsyncRenderHandler(ctx: DashboardContext, cmd: RequestAsyncRender): any {
    // eslint-disable-next-line no-console
    console.debug("handling request async render", cmd, "in context", ctx);

    try {
        const {
            payload: { id },
            correlationId,
        } = cmd;

        return asyncRenderRequested(id, ctx, correlationId);
    } catch (e) {
        throw internalErrorOccurred(
            ctx,
            "An unexpected error has occurred while requesting async rendering",
            e,
            cmd.correlationId,
        );
    }
}
