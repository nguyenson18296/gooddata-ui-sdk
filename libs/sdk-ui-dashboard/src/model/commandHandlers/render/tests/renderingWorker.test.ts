// (C) 2021 GoodData Corporation
import { loadDashboard } from "../../../commands";
import { DashboardTester } from "../../../tests/DashboardTester";
import { EmptyDashboardIdentifier } from "../../../tests/Dashboard.fixtures";
import { requestAsyncRender, resolveAsyncRender } from "../../../commands/render";

describe("renderingWorker", () => {
    let Tester: DashboardTester;

    describe("maximum timeout", () => {
        const asyncRenderRequestedTimeout = 200;
        const asyncRenderResolvedTimeout = 200;
        const maxTimeout = 400;

        beforeEach(
            () =>
                (Tester = DashboardTester.forRecording(EmptyDashboardIdentifier, {
                    renderingWorkerConfig: {
                        asyncRenderRequestedTimeout,
                        asyncRenderResolvedTimeout,
                        maxTimeout,
                    },
                })),
        );

        it("should emit render resolved, when async rendering is still running, but the maximum timeout reached", async () => {
            const componentId = "component";
            await Tester.waitFor("GDC.DASH/EVT.RENDER.REQUESTED");
            await Tester.dispatchAndWaitFor(loadDashboard(), "GDC.DASH/EVT.LOADED");
            Tester.dispatch(requestAsyncRender(componentId));
            await Tester.waitFor("GDC.DASH/EVT.RENDER.RESOLVED", maxTimeout);
            expect(Tester.emittedEventsDigest()).toMatchSnapshot();
        });
    });

    describe("async rendering resolution", () => {
        const asyncRenderRequestedTimeout = 200;
        const asyncRenderResolvedTimeout = 200;
        const maxTimeout = 2000;

        beforeEach(
            () =>
                (Tester = DashboardTester.forRecording(EmptyDashboardIdentifier, {
                    renderingWorkerConfig: {
                        asyncRenderRequestedTimeout,
                        asyncRenderResolvedTimeout,
                        maxTimeout,
                    },
                })),
        );

        it("should emit render resolved, when no async rendering is requested during the asyncRenderRequestedTimeout", async () => {
            await Tester.waitFor("GDC.DASH/EVT.RENDER.REQUESTED");
            await Tester.dispatchAndWaitFor(loadDashboard(), "GDC.DASH/EVT.LOADED");
            await Tester.waitFor("GDC.DASH/EVT.RENDER.RESOLVED", asyncRenderRequestedTimeout);
            expect(Tester.emittedEventsDigest()).toMatchSnapshot();
        });

        it("should emit render resolved, when async renderings are resolved during the asyncRenderRequestedTimeout", async () => {
            const componentId = "component";
            await Tester.waitFor("GDC.DASH/EVT.RENDER.REQUESTED");
            await Tester.dispatchAndWaitFor(loadDashboard(), "GDC.DASH/EVT.LOADED");
            Tester.dispatch(requestAsyncRender(componentId));
            Tester.dispatch(resolveAsyncRender(componentId));
            await Tester.waitFor("GDC.DASH/EVT.RENDER.RESOLVED", asyncRenderRequestedTimeout);
            expect(Tester.emittedEventsDigest()).toMatchSnapshot();
        });

        it("should emit render resolved after async rendering resolution", async () => {
            const componentId = "component";
            await Tester.waitFor("GDC.DASH/EVT.RENDER.REQUESTED");
            await Tester.dispatchAndWaitFor(loadDashboard(), "GDC.DASH/EVT.LOADED");
            Tester.dispatch(requestAsyncRender(componentId));
            await Tester.wait(asyncRenderRequestedTimeout);
            Tester.dispatch(resolveAsyncRender(componentId));
            await Tester.waitFor("GDC.DASH/EVT.RENDER.RESOLVED", asyncRenderResolvedTimeout);
            expect(Tester.emittedEventsDigest()).toMatchSnapshot();
        });

        it("should not emit render resolved, when async rendering is not resolved", async () => {
            const componentId = "component";
            await Tester.waitFor("GDC.DASH/EVT.RENDER.REQUESTED");
            await Tester.dispatchAndWaitFor(loadDashboard(), "GDC.DASH/EVT.LOADED");
            Tester.dispatch(requestAsyncRender(componentId));
            await Tester.wait(asyncRenderRequestedTimeout);
            await expect(
                Tester.waitFor("GDC.DASH/EVT.RENDER.RESOLVED", asyncRenderResolvedTimeout),
            ).rejects.toThrowErrorMatchingSnapshot();
            expect(Tester.emittedEventsDigest()).toMatchSnapshot();
        });

        it("should accept new async rendering requests, when running requests are not yet resolved", async () => {
            const componentId = "component";
            const componentId2 = "component2";
            const componentId3 = "component3";
            await Tester.waitFor("GDC.DASH/EVT.RENDER.REQUESTED");
            await Tester.dispatchAndWaitFor(loadDashboard(), "GDC.DASH/EVT.LOADED");
            Tester.dispatch(requestAsyncRender(componentId));
            await Tester.wait(asyncRenderRequestedTimeout);
            Tester.dispatch(requestAsyncRender(componentId2));
            Tester.dispatch(requestAsyncRender(componentId3));
            await Tester.wait(300);
            Tester.dispatch(resolveAsyncRender(componentId));
            Tester.dispatch(resolveAsyncRender(componentId2));
            Tester.dispatch(resolveAsyncRender(componentId3));
            await Tester.waitFor("GDC.DASH/EVT.RENDER.RESOLVED", asyncRenderResolvedTimeout);
            expect(Tester.emittedEventsDigest()).toMatchSnapshot();
        });

        it("should accept new async rendering request, when it's fired within time limit after resolution", async () => {
            const componentId = "component";
            await Tester.waitFor("GDC.DASH/EVT.RENDER.REQUESTED");
            await Tester.dispatchAndWaitFor(loadDashboard(), "GDC.DASH/EVT.LOADED");
            Tester.dispatch(requestAsyncRender(componentId));
            await Tester.wait(asyncRenderRequestedTimeout);
            Tester.dispatch(resolveAsyncRender(componentId));
            Tester.dispatch(requestAsyncRender(componentId));
            await Tester.wait(300);
            Tester.dispatch(resolveAsyncRender(componentId));
            await Tester.waitFor("GDC.DASH/EVT.RENDER.RESOLVED", asyncRenderResolvedTimeout);
            expect(Tester.emittedEventsDigest()).toMatchSnapshot();
        });
    });
});
