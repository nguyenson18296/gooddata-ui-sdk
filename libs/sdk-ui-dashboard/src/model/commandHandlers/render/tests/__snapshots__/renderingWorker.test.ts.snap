// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderingWorker async rendering resolution should accept new async rendering request, when it's fired within time limit after resolution 1`] = `
Array [
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.REQUESTED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.FILTER_CONTEXT.DATE_FILTER.VALIDATION.FAILED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.LOADED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.ASYNC.REQUESTED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.ASYNC.RESOLVED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.ASYNC.REQUESTED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.ASYNC.RESOLVED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.RESOLVED",
  },
]
`;

exports[`renderingWorker async rendering resolution should accept new async rendering requests, when running requests are not yet resolved 1`] = `
Array [
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.REQUESTED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.FILTER_CONTEXT.DATE_FILTER.VALIDATION.FAILED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.LOADED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.ASYNC.REQUESTED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.ASYNC.REQUESTED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.ASYNC.REQUESTED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.ASYNC.RESOLVED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.ASYNC.RESOLVED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.ASYNC.RESOLVED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.RESOLVED",
  },
]
`;

exports[`renderingWorker async rendering resolution should emit render resolved after async rendering resolution 1`] = `
Array [
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.REQUESTED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.FILTER_CONTEXT.DATE_FILTER.VALIDATION.FAILED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.LOADED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.ASYNC.REQUESTED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.ASYNC.RESOLVED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.RESOLVED",
  },
]
`;

exports[`renderingWorker async rendering resolution should emit render resolved, when async renderings are resolved during the asyncRenderRequestedTimeout 1`] = `
Array [
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.REQUESTED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.FILTER_CONTEXT.DATE_FILTER.VALIDATION.FAILED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.LOADED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.ASYNC.REQUESTED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.ASYNC.RESOLVED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.RESOLVED",
  },
]
`;

exports[`renderingWorker async rendering resolution should emit render resolved, when no async rendering is requested during the asyncRenderRequestedTimeout 1`] = `
Array [
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.REQUESTED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.FILTER_CONTEXT.DATE_FILTER.VALIDATION.FAILED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.LOADED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.RESOLVED",
  },
]
`;

exports[`renderingWorker async rendering resolution should not emit render resolved, when async rendering is not resolved 1`] = `"Wait for action 'GDC.DASH/EVT.RENDER.RESOLVED' timed out after 200ms"`;

exports[`renderingWorker async rendering resolution should not emit render resolved, when async rendering is not resolved 2`] = `
Array [
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.REQUESTED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.FILTER_CONTEXT.DATE_FILTER.VALIDATION.FAILED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.LOADED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.ASYNC.REQUESTED",
  },
]
`;

exports[`renderingWorker maximum timeout should emit render resolved, when async rendering is still running, but the maximum timeout reached 1`] = `
Array [
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.REQUESTED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.FILTER_CONTEXT.DATE_FILTER.VALIDATION.FAILED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.LOADED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.ASYNC.REQUESTED",
  },
  Object {
    "correlationId": undefined,
    "type": "GDC.DASH/EVT.RENDER.RESOLVED",
  },
]
`;
